File: examples/go/basic_use.go
//go:build ignore

package main

import "fmt"

// WITHOUT generics: separate types, duplicated code

// IntStack only works for int.
type IntStack struct{ items []int }

func (s *IntStack) Push(v int)        { s.items = append(s.items, v) }
func (s *IntStack) Len() int          { return len(s.items) }
func (s *IntStack) Pop() (int, bool) {
	if len(s.items) == 0 {
		return 0, false
	}
	top := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return top, true
}

// StringStack is an identical copy — just with string instead of int.
type StringStack struct{ items []string }

func (s *StringStack) Push(v string)        { s.items = append(s.items, v) }
func (s *StringStack) Len() int             { return len(s.items) }
func (s *StringStack) Pop() (string, bool) {
	if len(s.items) == 0 {
		return "", false
	}
	top := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return top, true
}

// AnyStack avoids the duplication but loses type safety:
// compiler cannot catch s.Push(42) on a "string stack" — only a runtime panic can.
type AnyStack struct{ items []any }

func (s *AnyStack) Push(v any)       { s.items = append(s.items, v) }
func (s *AnyStack) Len() int         { return len(s.items) }
func (s *AnyStack) Pop() (any, bool) {
	if len(s.items) == 0 {
		return nil, false
	}
	top := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return top, true
}


// WITH GENERICS: one implementation, full type safety

type Stack[T any] struct{ items []T }

func (s *Stack[T]) Push(v T) { s.items = append(s.items, v) }
func (s *Stack[T]) Len() int { return len(s.items) }
func (s *Stack[T]) Pop() (T, bool) {
	var zero T
	if len(s.items) == 0 {
		return zero, false
	}
	top := s.items[len(s.items)-1]
	s.items = s.items[:len(s.items)-1]
	return top, true
}

func main() {
	fmt.Println("=== Without generics ===")

	var intStack IntStack
	intStack.Push(1)
	intStack.Push(2)
	intStack.Push(3)
	if v, ok := intStack.Pop(); ok {
		fmt.Println("IntStack popped:", v)
	}

	var stringStack StringStack
	stringStack.Push("go")
	stringStack.Push("generics")
	if v, ok := stringStack.Pop(); ok {
		fmt.Println("StringStack popped:", v)
	}

	// AnyStack: both pushes compile fine, error only surfaces at runtime
	var anyStack AnyStack
	anyStack.Push(42)
	anyStack.Push("oops")
	if v, ok := anyStack.Pop(); ok {
		_ = v.(string) // runtime type assertion, panics if wrong
		fmt.Println("AnyStack top was a string: ", v)

		// This would panic at runtime:
		// _ = v.(int)
	}

	fmt.Println("\n=== With generics ===")

	var genericIntStack Stack[int]
	genericIntStack.Push(10)
	genericIntStack.Push(20)
	genericIntStack.Push(30)

	// This would not compile: cannot use string as int
	// genericIntStack.Push("hello")

	if v, ok := genericIntStack.Pop(); ok {
		fmt.Println("Stack[int] popped:", v)
	}

	var genericStringStack Stack[string]
	genericStringStack.Push("go")
	genericStringStack.Push("generics")

	// This would not compile: cannot use int as string
	// genericStringStack.Push(123)
	
	if v, ok := genericStringStack.Pop(); ok {
		fmt.Println("Stack[string] popped:", v)
	}
}

File: examples/typescript/basic_use.ts
export {};

// WITHOUT generics: separate types, duplicated code

// IntStack only works for number.
class IntStack {
  private items: number[] = [];

  push(v: number): void {
    this.items.push(v);
  }
  len(): number {
    return this.items.length;
  }
  pop(): [number, boolean] {
    if (this.items.length === 0) return [0, false];
    return [this.items.pop()!, true];
  }
}

// StringStack is an identical copy — just with string instead of number.
class StringStack {
  private items: string[] = [];

  push(v: string): void {
    this.items.push(v);
  }
  len(): number {
    return this.items.length;
  }
  pop(): [string, boolean] {
    if (this.items.length === 0) return ["", false];
    return [this.items.pop()!, true];
  }
}

// AnyStack avoids the duplication but loses type safety:
// compiler cannot catch push(42) on a "string stack" — only a runtime error can.
class AnyStack {
  private items: any[] = [];

  push(v: any): void {
    this.items.push(v);
  }
  len(): number {
    return this.items.length;
  }
  pop(): [any, boolean] {
    if (this.items.length === 0) return [undefined, false];
    return [this.items.pop(), true];
  }
}

// WITH GENERICS: one implementation, full type safety

class Stack<T> {
  private items: T[] = [];

  push(v: T): void {
    this.items.push(v);
  }
  len(): number {
    return this.items.length;
  }
  pop(): [T | undefined, boolean] {
    if (this.items.length === 0) return [undefined, false];
    return [this.items.pop()!, true];
  }
}

function main() {
  console.log("=== Without generics ===");

  const intStack = new IntStack();
  intStack.push(1);
  intStack.push(2);
  intStack.push(3);
  const [intVal, intOk] = intStack.pop();
  if (intOk) console.log("IntStack popped:", intVal);

  const stringStack = new StringStack();
  stringStack.push("go");
  stringStack.push("generics");
  const [strVal, strOk] = stringStack.pop();
  if (strOk) console.log("StringStack popped:", strVal);

  // AnyStack: both pushes compile fine, error only surfaces at runtime.
  const anyStack = new AnyStack();
  anyStack.push(42);
  anyStack.push("oops");
  const [anyVal, anyOk] = anyStack.pop();
  if (anyOk) {
    const s = anyVal as string; // runtime cast, throws if wrong
    console.log("AnyStack top was a string:", s);

    // This would throw at runtime:
    // const n = anyVal as number;
  }

  console.log("\n=== With generics ===");

  const genericIntStack = new Stack<number>();
  genericIntStack.push(10);
  genericIntStack.push(20);
  genericIntStack.push(30);

  // This would not compile: Argument of type 'string' is not assignable to parameter of type 'number'
  // genericIntStack.push("hello");

  const [gIntVal, gIntOk] = genericIntStack.pop();
  if (gIntOk) console.log("Stack<number> popped:", gIntVal);

  const genericStringStack = new Stack<string>();
  genericStringStack.push("go");
  genericStringStack.push("generics");

  // This would not compile: Argument of type 'number' is not assignable to parameter of type 'string'
  // genericStringStack.push(123);
  
  const [gStrVal, gStrOk] = genericStringStack.pop();
  if (gStrOk) console.log("Stack<string> popped:", gStrVal);
}

main();
